integrations/weaviate/src/haystack_integrations/document_stores/weaviate/document_store.py

```diff
@@ -216,6 +193,24 @@ def _to_data_object(self, document: Document) -> Dict[str, Any]:
 
         return data
 
+    def _convert_weaviate_v4_object_to_v3_object(self, data: Object) -> Dict[str, Any]:
```

Comment:
What's the need of this function? Can't we do it this in `_to_document()`?

Reply:
`_to_document()` assumes the result from a weaviate query is a dict. However, in v4, it is now an `Object`, so I wrote this function to do the conversion instead of modifying  `_to_document()`. But we could certainly do it in `_to_document()`.

Do you prefer the conversion to be done in `_to_document()` or update the sig so that `_to_document()` accepts `Object`? I'm not sure what impact the latter would have on the rest of the integration.

Reply:
Feel free to change it's signature if necessary, it's only meant to be used in the internals.

Ideally `_to_document()` should take whatever is returned by the Weaviate client and return a `Document`. If in v4 it's an `Object` that should be what it expects.

Reply:
I've evaluated the impact of changing `data` into `Object` and see that the change will be more complicated than it needs to be.

The reason is that this integration flattens documents before storing them in weaviate.

Since this is just an internal method, why don't we keep it like this until weaviate has full support for nested objects? We can include this change as part of a PR to update this part of the code:

https://github.com/deepset-ai/haystack-core-integrations/blob/a430cd630d15b9589c554d0e16d1cb7fac027a91/integrations/weaviate/src/haystack_integrations/document_stores/weaviate/_filters.py#L247-L251

Reply:
I don't get it, why is it more complex?

Right now you're calling with `_to_document()` passing the result of `_convert_weaviate_v4_object_to_v3_object()` to it. Where's the problem in making `_to_document()` take an `Object` and do the whole work of converting it to `Document`?

Reply:
you're right, it's actually easy. I thought I had to unflatten the metadata field too but turns out there's already another function to do that